src/webui/webapplication.cpp

```diff
@@ -48,114 +60,84 @@
-    ADD_ACTION(command, getGlobalUpLimit);
-    ADD_ACTION(command, getGlobalDlLimit);
-    ADD_ACTION(command, setGlobalUpLimit);
-    ADD_ACTION(command, setGlobalDlLimit);
-    ADD_ACTION(command, getTorrentsUpLimit);
-    ADD_ACTION(command, getTorrentsDlLimit);
-    ADD_ACTION(command, setTorrentsUpLimit);
-    ADD_ACTION(command, setTorrentsDlLimit);
-    ADD_ACTION(command, alternativeSpeedLimitsEnabled);
-    ADD_ACTION(command, toggleAlternativeSpeedLimits);
-    ADD_ACTION(command, toggleSequentialDownload);
-    ADD_ACTION(command, toggleFirstLastPiecePrio);
-    ADD_ACTION(command, setSuperSeeding);
-    ADD_ACTION(command, setForceStart);
-    ADD_ACTION(command, delete);
-    ADD_ACTION(command, deletePerm);
-    ADD_ACTION(command, increasePrio);
-    ADD_ACTION(command, decreasePrio);
-    ADD_ACTION(command, topPrio);
-    ADD_ACTION(command, bottomPrio);
-    ADD_ACTION(command, setLocation);
-    ADD_ACTION(command, rename);
-    ADD_ACTION(command, setAutoTMM);
-    ADD_ACTION(command, recheck);
-    ADD_ACTION(command, setCategory);
-    ADD_ACTION(command, addCategory);
-    ADD_ACTION(command, removeCategories);
-    ADD_ACTION(command, getSavePath);
-    ADD_ACTION(version, api);
-    ADD_ACTION(version, api_min);
-    ADD_ACTION(version, qbittorrent);
-
-    return actions;
-}
+constexpr Utils::Version<int, 3, 2> API_VERSION {2, 0, 0};
+constexpr int COMPAT_API_VERSION = 18;
+constexpr int COMPAT_API_VERSION_MIN = 18;
+
+constexpr int MAX_ALLOWED_FILESIZE = 10 * 1024 * 1024;
+constexpr int BAN_TIME = 3600000; // 1 hour
+constexpr int MAX_AUTH_FAILED_ATTEMPTS = 5;
+
+const QString PATH_PREFIX_API {"/api/"};
+const QString PATH_PREFIX_IMAGES {"/images/"};
+const QString PATH_PREFIX_THEME {"/theme/"};
+const QString WWW_FOLDER {":/www"};
+const QString PUBLIC_FOLDER {"/public"};
+const QString PRIVATE_FOLDER {"/private"};
+const QString MAX_AGE_MONTH {"public, max-age=2592000"};
+
+// WebSession
+
+struct WebSession
+{
+    const QString id;
+    uint timestamp;
+    WebSessionData data;
+
+    WebSession(const QString& id)
+        : id(id)
+    {
+        updateTimestamp();
+    }
+
+    void updateTimestamp()
+    {
+        timestamp = QDateTime::currentDateTime().toTime_t();
+    }
+};
 
 namespace
 {
-#define CHECK_URI(ARGS_NUM) \
-    if (args_.size() != ARGS_NUM) { \
-        status(404, "Not Found"); \
-        return; \
-    }
-#define CHECK_PARAMETERS(PARAMETERS) \
-    QStringList parameters; \
-    parameters << PARAMETERS; \
-    if (parameters.size() != request().posts.size()) { \
-        status(400, "Bad Request"); \
-        return; \
-    } \
-    foreach (QString key, request().posts.keys()) { \
-        if (!parameters.contains(key, Qt::CaseInsensitive)) { \
-            status(400, "Bad Request"); \
-            return; \
-        } \
+    class BadRequestError : public std::runtime_error
+    {
+    public:
+        explicit BadRequestError(const QString &message = "")
```

Comment:
can't tell why, but I expected to see `std::string` as parameter.

Reply:
I see no reason to do it.
Ths is a class declared/defined in Qt application, it will be used together with other Qt types include QString as well. The fact it extends std class and uses std::string internally shouldn't forbid us to have convenient interface with all Qt world outside it.